### PR TITLE
refactor: extension provider

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
1. abstract public extension provider
2. adjust pipeline and dicehub extension provider
3. fix latest extension version was overwrite due to service reboot

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/gantt?id=343231&iterationID=1440&pId=0&type=REQUIREMENT)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |     1. abstract public extension provider 2. adjust pipeline and dicehub extension provider         |
| 🇨🇳 中文    |    抽象并调整dicehub和pipeline的extension逻辑          |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: api/proto/core/dicehub/extension/extension.proto [hunk 1]
```diff
@@ -165,6 +165,7 @@ message ExtensionVersionCreateRequest {
     bool forceUpdate = 8;
     bool all = 9;
     bool isDefault = 10;
+    optional google.protobuf.Timestamp updateAt = 11;
```
Comment (sfwn) [inline api/proto/core/dicehub/extension/extension.proto:168]: we normally use `updatedAt`

File: cmd/pipeline/bootstrap.yaml [hunk 2]
```diff
@@ -63,10 +63,21 @@ edge-reporter:
   target:
     url: "${OPENAPI_PUBLIC_URL}"
     auth_token: "${OPENAPI_TOKEN}"
-
+erda.apps.gallery: { }
```
Comment (sfwn) [inline cmd/pipeline/bootstrap.yaml:66]: reverse depends it not permitted.

File: internal/apps/dop/dicehub/release/release.service.go [hunk 3]
```diff
@@ -58,7 +58,7 @@ type ReleaseService struct {
 	db              *db.ReleaseConfigDB
 	labelRelationDB *db.LabelRelationConfigDB
 	imageDB         *imagedb.ImageConfigDB
-	extensionDB     *extensiondb.ExtensionConfigDB
+	extensionDB     *db2.Client
```
Comment (sfwn) [inline internal/apps/dop/dicehub/release/release.service.go:61]: bad db2

File: cmd/erda-server/bootstrap.yaml [hunk 4]
```diff
@@ -81,9 +81,11 @@ erda.core.dicehub.image: { }
 erda.core.dicehub.release:
   max_time_reserved: "${RELEASE_MAX_TIME_RESERVED:72}"
   gc_switch: "${RELEASE_GC_SWITCH:true}"
-erda.core.dicehub.extension:
+extension:
```
Comment (sfwn) [inline cmd/erda-server/bootstrap.yaml:84]: why there are two extension providers?

**Review by sfwn — CHANGES_REQUESTED** [review]

File: api/proto/core/dicehub/extension/extension.proto (hunk 1 above)
Comment (chengjoey) [reply api/proto/core/dicehub/extension/extension.proto]: adjusted

File: cmd/pipeline/bootstrap.yaml (hunk 2 above)
Comment (chengjoey) [reply cmd/pipeline/bootstrap.yaml]: adjusted, gallery listen extension event

File: internal/apps/dop/dicehub/release/release.service.go (hunk 3 above)
Comment (chengjoey) [reply internal/apps/dop/dicehub/release/release.service.go]: adjusted

File: cmd/erda-server/bootstrap.yaml (hunk 4 above)
Comment (chengjoey) [reply cmd/erda-server/bootstrap.yaml:84]: `extension` is new public provider

File: api/proto/apps/gallery/gallery.proto [hunk 5]
```diff
@@ -51,6 +52,15 @@ service Gallery {
     };
   }
 
+  rpc ExtensionHook(common.EventCreateRequest) returns(common.EventCreateResponse) {
+    option(google.api.http) = {
+      post: "/api/extension-hook",
```
Comment (sfwn) [inline api/proto/apps/gallery/gallery.proto:57]: `/api/hooks/actions/gallery-extension-hook`

File: api/proto/common/event.proto [hunk 6]
```diff
@@ -0,0 +1,22 @@
+syntax = "proto3";
+
+package erda.common;
+
+import "google/protobuf/struct.proto";
+
+option go_package = "github.com/erda-project/erda-proto-go/common/pb";
+
+message EventCreateRequest {
```
Comment (sfwn) [inline api/proto/common/event.proto:9]: remove `apistructs.EventCreateRequest` later

File: apistructs/header.go [hunk 7]
```diff
@@ -14,6 +14,16 @@
 
 package apistructs
 
+var (
+	EventActionCreate = "create"
```
Comment (sfwn) [inline apistructs/header.go:18]: Bad location at `header.go`. You should put nearby `extension`.

Comment (sfwn) [reply apistructs/header.go:18]: And can you use proto enum?

File: cmd/erda-server/bootstrap.yaml (hunk 4 above)
Comment (sfwn) [reply cmd/erda-server/bootstrap.yaml:84]: Unclear. You should declare the difference between `extension` and `erda.core.dicehub.extension`.

File: cmd/erda-server/bootstrap.yaml [hunk 8]
```diff
@@ -81,9 +81,11 @@ erda.core.dicehub.image: { }
...
   extension_sources: "${EXTENSION_SOURCES}"
   extension_sources_cron: "${EXTENSION_SOURCES_CRON:0 */5 * * * ?}"
+  reload_extension_type: "${RELOAD_EXTENSION_TYPE:addon}"
+erda.core.dicehub.extension: { }
```
Comment (sfwn) [inline cmd/erda-server/bootstrap.yaml:88]: I think there is no `dicehub.extension` anymore after this pr.

File: internal/apps/dop/services/apierrors/errors.go [hunk 9]
```diff
@@ -567,6 +567,7 @@ var (
 	ErrQueryExtension                  = err("ErrQueryExtension", "查询扩展失败")
 	ErrCreateExtensionVersion          = err("ErrCreateExtensionVersion", "添加扩展版本失败")
 	ErrQueryExtensionVersion           = err("ErrQueryExtensionVersion", "查询扩展版本失败")
+	ErrDeleteExtensionVersion          = err("ErrDeleteExtensionVersion", "删除扩展版本失败")
```
Comment (sfwn) [inline internal/apps/dop/services/apierrors/errors.go:570]: cannot be used at pkg level

File: internal/tools/pipeline/providers/reconciler/taskrun/taskop/prepare.go [hunk 10]
```diff
@@ -191,7 +191,7 @@ func (pre *prepare) makeTaskRun() (needRetry bool, err error) {
 	extSearchReq := make([]string, 0)
 	extSearchReq = append(extSearchReq, getActionAgentTypeVersion())
 	extSearchReq = append(extSearchReq, pre.ActionMgr.MakeActionTypeVersion(&task.Extra.Action))
-	actionDiceYmlJobMap, actionSpecYmlJobMap, err := pre.ActionMgr.SearchActions(extSearchReq, pre.ActionMgr.MakeActionLocationsBySource(p.PipelineSource),
```
Comment (sfwn) [inline internal/tools/pipeline/providers/reconciler/taskrun/taskop/prepare.go:194]: why remove it

File: cmd/erda-server/bootstrap.yaml (hunk 8 above)
Comment (sfwn) [reply cmd/erda-server/bootstrap.yaml:88]: move `dicehub.extension` to `internal/pkg/extension`.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/tools/pipeline/providers/reconciler/taskrun/taskop/prepare.go (hunk 10 above)
Comment (chengjoey) [reply internal/tools/pipeline/providers/reconciler/taskrun/taskop/prepare.go:194]: location is not existed in extension

File: api/proto/apps/gallery/gallery.proto (hunk 5 above)
Comment (chengjoey) [reply api/proto/apps/gallery/gallery.proto:57]: adjusted

File: apistructs/header.go (hunk 7 above)
Comment (chengjoey) [reply apistructs/header.go:18]: moved to apistructs/extension.go

Comment (chengjoey) [reply apistructs/header.go:18]: > And can you use proto enum?

enum field is int32 kind, but request event by bundle is string type

File: cmd/erda-server/bootstrap.yaml (hunk 8 above)
Comment (chengjoey) [reply cmd/erda-server/bootstrap.yaml:88]: adjusted

File: internal/apps/dop/services/apierrors/errors.go (hunk 9 above)
Comment (chengjoey) [reply internal/apps/dop/services/apierrors/errors.go:570]: adjusted, moved to extension pkg

File: cmd/erda-server/main.go [hunk 11]
```diff
@@ -204,12 +203,12 @@ import (
 	_ "github.com/erda-project/erda/internal/apps/msp/resource/deploy/handlers/zookeeper"
 	_ "github.com/erda-project/erda/internal/apps/msp/tenant"
 	_ "github.com/erda-project/erda/internal/apps/msp/tenant/project"
+	_ "github.com/erda-project/erda/internal/pkg/extension"
 	_ "github.com/erda-project/erda/internal/tools/monitor/core/settings/retention-strategy"
 	_ "github.com/erda-project/erda/internal/tools/monitor/core/storekit/elasticsearch/index/cleaner"
 	_ "github.com/erda-project/erda/internal/tools/monitor/core/storekit/elasticsearch/index/loader"
 	_ "github.com/erda-project/erda/internal/tools/monitor/core/storekit/elasticsearch/index/retention-strategy"
 	"github.com/erda-project/erda/internal/tools/monitor/extensions/loghub"
-	_ "github.com/erda-project/erda/internal/tools/monitor/extensions/loghub/index/query"
```
Comment (sfwn) [inline cmd/erda-server/main.go:212]: needed

File: cmd/erda-server/bootstrap.yaml [hunk 12]
```diff
@@ -81,9 +81,11 @@ erda.core.dicehub.image: { }
 erda.core.dicehub.release:
   max_time_reserved: "${RELEASE_MAX_TIME_RESERVED:72}"
   gc_switch: "${RELEASE_GC_SWITCH:true}"
-erda.core.dicehub.extension:
+erda.core.extension:
   extension_sources: "${EXTENSION_SOURCES}"
   extension_sources_cron: "${EXTENSION_SOURCES_CRON:0 */5 * * * ?}"
+  reload_extension_type: "${RELOAD_EXTENSION_TYPE:addon}"
+  enable_service: "${ENABLE_SERVICE:true}"
```
Comment (sfwn) [inline cmd/erda-server/bootstrap.yaml:88]: ERDA_CORE_EXTENSION_ENABLE_SERVICE

**Review by sfwn — CHANGES_REQUESTED** [review]

File: cmd/erda-server/main.go (hunk 11 above)
Comment (chengjoey) [reply cmd/erda-server/main.go]: adjusted

File: cmd/erda-server/bootstrap.yaml (hunk 12 above)
Comment (chengjoey) [reply cmd/erda-server/bootstrap.yaml]: adjusted

**Review by sfwn — APPROVED** [review]

**Pull request merged**